docs/IPC.md

```diff
@@ -35,3 +76,206 @@ Each subscriber's library:
+
+Upon publishing changes, the [Publication](https://pkg.go.dev/github.com/lf-edge/eve/pkg/pillar@v0.0.0-20220603153046-23f5ce4eb5ee/pubsub#Publication) is responsible for:
+
+1. [Validating that the topic and types fit the `Publication`](https://github.com/lf-edge/eve/blob/23f5ce4eb5ee/pkg/pillar/pubsub/publish.go#L100-L110)
+1. [Saving the data in memory](https://github.com/lf-edge/eve/blob/23f5ce4eb5ee/pkg/pillar/pubsub/publish.go#L134)
+1. [Marshaling the go object into json bytes](https://github.com/lf-edge/eve/blob/23f5ce4eb5ee/pkg/pillar/pubsub/publish.go#L141).
+1. [Updating the driver with the key and updated bytes](https://github.com/lf-edge/eve/blob/23f5ce4eb5ee/pkg/pillar/pubsub/publish.go#L148)
+
+Note that there is nothing about notifications, subscriptions, or persistence. The `Publication` is solely responsible for
+validation, storing in memory, marshaling to json, and updating the `Driver` with the json bytes.
+
+It is the `Driver` that handles notification and persistence.
+
+When the `Driver` receives the update, it:
+
+1. Stores any persistent data, using whatever storage is appropriate for the driver.
+1. Notifies any subscribers of changes, using whatever notification mechanism is appropriate for the driver.
+
+All persistence and notification - the publish-subscribe part of pubsub - happens entirely within the driver.
+The driver could be memory, network communication, socket communications, polled files, anything at all. The
+`Publication` does not care.
+
+The publishing process can retrieve data using `Get()`, `GetAll()` and `Iterate()`. These work entirely on the local
+in-memory copy of the `Publisher`.
+
+### Subscriber
+
+When a process wants to retrieve information stored by a publishing process, it does the following:
+
+1. Use the `PubSub` factory to [create a new Subscription](https://pkg.go.dev/github.com/lf-edge/eve/pkg/pillar@v0.0.0-20220603153046-23f5ce4eb5ee/pubsub#PubSub.NewSubscription), passing it the `AgentName`, `AgentScope`, and `TopicType`. In addition, pass it handlers that should be called
+for modifications of data, such as creating a new entry, updating an entry, or deleting an entry.
+1. Use the returned [Subscription](https://pkg.go.dev/github.com/lf-edge/eve/pkg/pillar@v0.0.0-20220603153046-23f5ce4eb5ee/pubsub#Subscription) to and retrieve objects based on keys.
+
+The [Subscription](https://pkg.go.dev/github.com/lf-edge/eve/pkg/pillar@v0.0.0-20220603153046-23f5ce4eb5ee/pubsub#Subscription) is responsible for:
+
+1. Being informed of updates from the driver.
+1. Unmarshaling the raw json bytes into the correct object type.
+1. Saving the data to its own in-memory copy of the table.
+1. Calling appropriate handlers for state changes.
+
+Note that there is nothing about notifications, publications, or persistence. The `Subscription` is solely responsible for
+receiving updates from the `Driver`, validation, storing in memory, and calling handlers.
```

Comment:
Probably we should mention here that we have slightly different logic for persist subscription: during Activate() we call populate function [here](https://github.com/lf-edge/eve/blob/dee0c391e23f40e08b7e7eacf2e532c03086c846/pkg/pillar/pubsub/subscribe.go#L54). Please see the comment https://github.com/lf-edge/eve/blob/dee0c391e23f40e08b7e7eacf2e532c03086c846/pkg/pillar/pubsub/subscribe.go#L77-L79

Reply:
Good point. I had forgotten about that. I am adding it now.

We also should add info about the restart counter and what it is for, but I think that can wait for a future PR.

Reply:
Updated, thank you @giggsoff